### PR TITLE
chore: adds fs access check at startup time

### PR DESCRIPTION
**Pull request by M4tteoP**
Fix https://github.com/corazawaf/coraza/issues/1007

### CODE REVIEW

File: internal/environment/checks.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"fmt"
+	"os"
+
+	"github.com/corazawaf/coraza/v3/debuglog"
+)
+
+// CheckFSAccess is a helper function to check if the WAF has access to the filesystem
+func CheckFSAccess(logger debuglog.Logger, tmpDir string) error {
+	if HasAccessToFS {
```
Comment (jcchavezs) [inline internal/environment/checks.go:15]: I'd invert the conditional to avoid indentation.

File: internal/environment/checks.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+		file, err := os.CreateTemp(tmpDir, "checkfsfile")
+		if err != nil {
+			return fmt.Errorf("create file: %w", err)
+		}
+		defer file.Close()
+
+		_, err = file.WriteString("check fs data")
```
Comment (jcchavezs) [inline internal/environment/checks.go:22]: Not sure this check is 100% needed as CreateTemp is creating a file and opening it for writing hence not sure about the need of actually writing in it, also if we need that I'd write a single byte as that seems to be enough cc @anuraaga

File: internal/environment/checks.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		if err != nil {
+			return fmt.Errorf("write to file: %w", err)
+		}
+
+		err = os.Remove(file.Name())
```
Comment (jcchavezs) [inline internal/environment/checks.go:27]: not sure about this. Here we are removing the file but lines before we defer on closing. Also, I think if we truly remove, a file and forget is enough.

Comment (M4tteoP) [reply internal/environment/checks.go]: Thanks, moved into the defer function as a fire and forget call

File: internal/environment/checks.go (hunk 2 above)
Comment (M4tteoP) [reply internal/environment/checks.go]: Thanks, I took a better look at CreateTemp, and your thoughts are reasonable. I've removed it for now

File: internal/environment/checks.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import (
+	"fmt"
+	"os"
+
+	"github.com/corazawaf/coraza/v3/debuglog"
+)
+
+// IsDirWritable is a helper function to check if the WAF has access to the filesystem
+func IsDirWritable(logger debuglog.Logger, dir string) error {
+	if !HasAccessToFS {
```
Comment (jcchavezs) [inline internal/environment/checks.go:15]: I'd rather decouple this from the other as we are mixing two concerns here. Also it is possible that for `tinygo.wasm` we just return false.

File: internal/environment/checks.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+	}
+	file, err := os.CreateTemp(dir, "checkfsfile")
+	if err != nil {
+		return fmt.Errorf("create file: %w", err)
+	}
+	defer func() {
+		file.Close()
+		os.Remove(file.Name())
+	}()
+	logger.Debug().Msg("Filesystem access check successful")
```
Comment (jcchavezs) [inline internal/environment/checks.go:27]: I think it is enough to drop this.

File: internal/environment/checks.go (hunk 4 above)
Comment (M4tteoP) [reply internal/environment/checks.go]: I split the function based on the build tag and wrapped it checking HasAccessToFS. PTAL

File: internal/environment/default.go [hunk 6]
```diff
@@ -6,6 +6,26 @@
...
+
+	"github.com/corazawaf/coraza/v3/debuglog"
+)
+
 // HasAccessToFS indicates whether the runtime target environment has access
 // to OS' filesystem or not.
 var HasAccessToFS = true
+
+// IsDirWritable is a helper function to check if the WAF has access to the filesystem
+func IsDirWritable(logger debuglog.Logger, dir string) error {
```
Comment (jcchavezs) [inline internal/environment/default.go:21]: Unless I am missing something I think logger isn't needed here.

File: internal/environment/nofsaccess.go [hunk 7]
```diff
@@ -6,4 +6,12 @@
...
 package environment
 
+import "github.com/corazawaf/coraza/v3/debuglog"
+
 var HasAccessToFS = false
+
+// IsDirWritable is a helper function to check if the WAF has access to the filesystem
+// It is unexpected to call this function when no_fs_access build tag is enabled
+func IsDirWritable(logger debuglog.Logger, dir string) error {
+	panic("Unexpected call to IsDirWritable with no_fs_access build tag")
```
Comment (jcchavezs) [inline internal/environment/nofsaccess.go:16]: cc @anuraaga

File: internal/environment/default.go (hunk 6 above)
Comment (M4tteoP) [reply internal/environment/default.go]: Thanks, it was a leftover

**Comment by jcchavezs** [comment]
Any input on this @anuraaga otherwise I think we can merge it.

**Review by anuraaga — APPROVED** [review]

**Pull request merged**